core/commonMain/src/kotlinx/serialization/encoding/ReorderingCompositeEncoder.kt

```diff
@@ -0,0 +1,240 @@
+    }
+
+    override fun endStructure(descriptor: SerialDescriptor) {
+        encodeBufferedFields(descriptor)
+    }
+
+    private fun encodeBufferedFields(descriptor: SerialDescriptor) {
+        val classEncoder = beginStructure(descriptor)
+        bufferedCalls.forEach { fieldToEncode ->
+            fieldToEncode?.encoder?.invoke(classEncoder)
+        }
+        classEncoder.endStructure(descriptor)
+    }
+
+    override fun encodeBooleanElement(descriptor: SerialDescriptor, index: Int, value: Boolean) {
+        bufferEncoding(descriptor, index) { encodeBooleanElement(descriptor, index, value) }
+    }
+
+    override fun encodeByteElement(descriptor: SerialDescriptor, index: Int, value: Byte) {
+        bufferEncoding(descriptor, index) { encodeByteElement(descriptor, index, value) }
+    }
+
+    override fun encodeCharElement(descriptor: SerialDescriptor, index: Int, value: Char) {
+        bufferEncoding(descriptor, index) { encodeCharElement(descriptor, index, value) }
+    }
+
+    override fun encodeDoubleElement(descriptor: SerialDescriptor, index: Int, value: Double) {
+        bufferEncoding(descriptor, index) { encodeDoubleElement(descriptor, index, value) }
+    }
+
+    override fun encodeFloatElement(descriptor: SerialDescriptor, index: Int, value: Float) {
+        bufferEncoding(descriptor, index) { encodeFloatElement(descriptor, index, value) }
+    }
+
+    override fun encodeIntElement(descriptor: SerialDescriptor, index: Int, value: Int) {
+        bufferEncoding(descriptor, index) { encodeIntElement(descriptor, index, value) }
+    }
+
+    override fun encodeLongElement(descriptor: SerialDescriptor, index: Int, value: Long) {
+        bufferEncoding(descriptor, index) { encodeLongElement(descriptor, index, value) }
+    }
+
+    override fun <T : Any> encodeNullableSerializableElement(
+        descriptor: SerialDescriptor,
+        index: Int,
+        serializer: SerializationStrategy<T>,
+        value: T?
+    ) {
+        bufferEncoding(descriptor, index) { encodeNullableSerializableElement(descriptor, index, serializer, value) }
+    }
+
+    override fun <T> encodeSerializableElement(
+        descriptor: SerialDescriptor,
+        index: Int,
+        serializer: SerializationStrategy<T>,
+        value: T
+    ) {
+        bufferEncoding(descriptor, index) { encodeSerializableElement(descriptor, index, serializer, value) }
+    }
+
+    override fun encodeShortElement(descriptor: SerialDescriptor, index: Int, value: Short) {
+        bufferEncoding(descriptor, index) { encodeShortElement(descriptor, index, value) }
+    }
+
+    override fun encodeStringElement(descriptor: SerialDescriptor, index: Int, value: String) {
+        bufferEncoding(descriptor, index) { encodeStringElement(descriptor, index, value) }
+    }
+
+    override fun encodeInlineElement(descriptor: SerialDescriptor, index: Int): Encoder {
+        return BufferingInlineEncoder(descriptor, index)
+    }
```

Comment:
Note that the implementation of this should be equivalent to:
```thisEncoder. encodeSerializableElement(MyData. serializer. descriptor, 0, MyInt. serializer(), myInt)```

Including the support of nested structs (nested structs always go through `encodeSerializableElement`)

Reply:
Yes, this is working like this after reordering, or maybe I misunderstood

Reply:
What about:
```kotlin
@Serializable data class Outer(val a: Middle, val b: Int)

@Serializable value class Middle(val i: Inner)

@Serializable value class Inner(val c: String, d: Boolean)
```
The way I read it, this at least can be implemented using `encodeInlineElement(d, 0)` on the composite encoder for outer. Which would then mean that the inner value is serialized by calling `encodeSerializableValue()` on that (`BufferingInlineEncoder`) encoder.

Reply:
Test has been successful with your example, so I still don't get your point https://github.com/Kotlin/kotlinx.serialization/pull/2722/files#r1648164815

> the inner value is serialized by calling encodeSerializableValue() on that (BufferingInlineEncoder) encoder
No, here are the steps:
- `encodeInlineElement(d, 0)` -> returns `BufferingInlineEncoder` with descriptor of `Middle` and index 0, but does not buffer anything for the moment (this is the role of `BufferingInlineEncoder`)
- `BufferingInlineEncoder.encodeSerializableValue` is called, so nothing more is done, no nested call is done even for `Inner` thanks to the following `bufferEncoding` buffering `compositeEncoderDelegate.encodeInlineElement(descriptor, elementIndex).encodeSerializableValue(serializer, value)`
- then when `endStructure` is called, and replays the elements encoding